R/model-tree.R

```diff
@@ -666,16 +724,79 @@ color_tree_by <- function(tree_data, color_by = "run"){
         )
       )
     }else{
-      # Gradient coloring: get colors for unique values (excluding NA)
+      # Gradient coloring; doesn't need to be sorted
       pal_bbr <- scales::pal_gradient_n(bbr_cols)(seq(0, 1, length.out = n_levels))
       tree_data$col <- factor(tree_data$col)
       levels(tree_data$col) <- c(node_colors, pal_bbr)
     }
+    # Set color_by attribute to color_by argument
+    attr(tree_data, "color_by") <- color_by
   }else{
     # all bbr red color
     pal_bbr <- scales::pal_gradient_n(bbr_cols)(1)
     tree_data$col <- factor(tree_data$col)
     levels(tree_data$col) <- c(node_colors, pal_bbr)
+    # Set color_by attribute to leafCount (default)
+    attr(tree_data, "color_by") <- "leafCount"
+  }
+
+  return(tree_data)
+}
+
+#' Create a size column based on the unique _numeric_ (or integer) values of
+#' another column.
+#' @inheritParams make_tree_tooltip
+#' @inheritParams model_tree
+#' @param rescale_to A numeric vector of length 2 specifying the range to rescale
+#'  `size_by` values to. Defaults to `c(1, 3)`, where `1` is the smallest node
+#'  size and `3` is the largest.
+#' @noRd
+size_tree_by <- function(tree_data, size_by = NULL, rescale_to = c(1, 3)){
+  if(!is.null(size_by)){
+    checkmate::assert_true(size_by %in% names(tree_data))
+    checkmate::assert_numeric(rescale_to, len = 2, lower = 1)
+
+    tree_data$node_size <- tree_data[[size_by]]
+
+    # Scale size with numeric value
+    if(inherits(tree_data$node_size, c("numeric", "integer"))){
+      # Rescale values based on standard deviation
+      # - a large SD can lead to very large nodes
+      mean_val <- mean(tree_data$node_size, na.rm = TRUE)
+      sd_val <- sd(tree_data$node_size, na.rm = TRUE)
+
+      if(sd_val != 0){
+        tree_data$node_size <- (tree_data$node_size - mean_val) / sd_val
+      }else{
+        # Would only happen if all values are the same
+        # - avoids dividing by 0
+        tree_data$node_size <- rep(1, length(tree_data$node_size))
+      }
```

Comment:
Are you sure any of this is needed?

```R
xs <- rnorm(100, mean = -30, sd = 70)
all.equal(scales::rescale(xs, to = c(0, 3)), scales::rescale((xs - mean(xs)) / sd(xs), to = c(0, 3)))
#> [1] TRUE
xs <- sample(1:1000, 100)
all.equal(scales::rescale(xs, to = c(0, 3)), scales::rescale((xs - mean(xs)) / sd(xs), to = c(0, 3)))
#> [1] TRUE
```

Reply:
I asked myself the same question and had trouble coming up with good examples to test but your example proves it to me. I Noticed that passing `rescale_to = c(1, 100)` made the size differences incredibly large and was trying to normalize to the standard deviation to avoid that, but I seem to have had a brain fart.